tools/yulPhaser/Program.cpp

```diff
@@ -0,0 +1,152 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <tools/yulPhaser/Program.h>
+
+#include <tools/yulPhaser/Exceptions.h>
+
+#include <liblangutil/CharStream.h>
+#include <liblangutil/ErrorReporter.h>
+#include <liblangutil/Exceptions.h>
+
+#include <libyul/AsmAnalysis.h>
+#include <libyul/AsmAnalysisInfo.h>
+#include <libyul/AsmJsonConverter.h>
+#include <libyul/AsmParser.h>
+#include <libyul/AsmPrinter.h>
+#include <libyul/YulString.h>
+#include <libyul/backends/evm/EVMDialect.h>
+#include <libyul/optimiser/Disambiguator.h>
+#include <libyul/optimiser/ForLoopInitRewriter.h>
+#include <libyul/optimiser/FunctionGrouper.h>
+#include <libyul/optimiser/FunctionHoister.h>
+#include <libyul/optimiser/Metrics.h>
+#include <libyul/optimiser/OptimiserStep.h>
+#include <libyul/optimiser/Suite.h>
+
+#include <libsolutil/JSON.h>
+
+#include <cassert>
+#include <memory>
+
+using namespace std;
+using namespace solidity;
+using namespace solidity::langutil;
+using namespace solidity::yul;
+using namespace solidity::util;
+
+namespace solidity::phaser
+{
+
+Program Program::load(CharStream& _sourceCode)
+{
+	// ASSUMPTION: parseSource() rewinds the stream on its own
+	Dialect const& dialect = EVMDialect::strictAssemblyForEVMObjects(EVMVersion{});
+	unique_ptr<Block> ast = parseSource(dialect, _sourceCode);
+	unique_ptr<AsmAnalysisInfo> analysisInfo = analyzeAST(dialect, *ast);
+
+	Program program(
+		dialect,
+		disambiguateAST(dialect, *ast, *analysisInfo)
+	);
+	program.optimise({
+		FunctionHoister::name,
+		FunctionGrouper::name,
+		ForLoopInitRewriter::name,
+	});
+
+	return program;
+}
+
+void Program::optimise(vector<string> const& _optimisationSteps)
+{
+	applyOptimisationSteps(m_dialect, m_nameDispenser, *m_ast, _optimisationSteps);
+}
+
+ostream& operator<<(std::ostream& _stream, Program const& _program)
+{
+	return _stream << AsmPrinter()(*_program.m_ast);
+}
+
+string Program::toJson() const
```

Comment:
Do we really need that?

Reply:
Not really. But it's very handy for inspecting the Yul code when writing tests or debugging. I could just keep it stashed constantly and apply it when I need it but I thought that it wouldn't hurt to leave it in in case it comes in handy later (it probably will). It's very short so it should not be big maintenance burden.

Reply:
Isn't the AsmPrinter much more useful?

Reply:
Depends. Both were useful in different situations. When writing tests I found it very helpful to see the exact AST structure, including the names of the classes used as its nodes. I could figure this out by looking at yul code and then finding the right definitions in `AsmData.h` but it would have taken me more time. And even then I could easily get it wrong - I think it's a good thing to be able to see exactly what's in the tree rather than assume based on my (more or less complete) understanding of it.

Reply:
Ah ok, makes sense given that in your tests you were actually checking the exact structure of the ast instead of just its pretty-printed form.